### ISSUE TITLE
Upgrade gets blocked if DSCI object name changes

### ISSUE THREAD

**Issue by bdattoma** [opendatahub-io/opendatahub-operator#718]
**Describe the bug**
While upgrade RHODS from 2.2.0 to 2.4.0, the process got blocked because DSCI name is different in the new version. 

After unblocking the upgrade by deleting the new DSCI, the process gets blocked again after the existing DSC gets updated to include the new components:

```
Component removal failed:
DSCInitialization.dscinitialization.opendatahub.io "rhods-setup" not
found
``` 

**To Reproduce**

1. Install RHODS (ODH) v2.2.0
2. Upgrade RHODS (ODH) v2.4.0
3. Check the DSCI of the operator after upgrade is triggered
4. Delete new DSCI in favor of old one

**Expected behavior**
If DSCI object name must change, operator must handle the migration

**Screenshots**

**Additional context**
Operator pod gets restarted and prints the following errors in the logs:
```
2023-11-09T15:02:44Z	INFO	controllers.DSCInitialization	Reconciling DSCInitialization.	{"DSCInitialization": "", "Request.Name": "rhods-setup"}
2023-11-09T15:02:44Z	ERROR	Reconciler error	{"controller": "dscinitialization", "controllerGroup": "dscinitialization.opendatahub.io", "controllerKind": "DSCInitialization", "DSCInitialization": {"name":"rhods-setup"}, "namespace": "", "name": "rhods-setup", "reconcileID": "aa8b0c90-c543-4e1a-8f3e-6022aef87272", "error": "only one instance of DSCInitialization object is allowed. Update existing instance on namespace  and name rhods-setup"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/remote-source/operator/deps/gomod/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:329
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/remote-source/operator/deps/gomod/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/remote-source/operator/deps/gomod/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235
2023-11-09T15:02:44Z	INFO	controllers.DataScienceCluster	Reconciling DataScienceCluster resources	{"Request.Namespace": "", "Request.Name": "default"}
2023-11-09T15:02:44Z	ERROR	Reconciler error	{"controller": "datasciencecluster", "controllerGroup": "datasciencecluster.opendatahub.io", "controllerKind": "DataScienceCluster", "DataScienceCluster": {"name":"default"}, "namespace": "", "name": "default", "reconcileID": "b9db1aec-1145-4858-9c8c-adb1c55b7d74", "error": "only one instance of DSCInitialization object is allowed"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/remote-source/operator/deps/gomod/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:329
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/remote-source/operator/deps/gomod/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/remote-source/operator/deps/gomod/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235
```

I think we could get siimilar issue with DSC name upgrade. I see that for this Release that name changed as well, but didn't compromise directly the upgrade. However, the next upgrade to a newer version should not have the expected name implicitly coded, but should infer it (maybe it already does)

**Comment by bdattoma** [comment]
After trying unblocking the process manually (deleting old DSCI in favor of new one), I bumped into new error about changing immutable fields:

```
Component reconciliation failed: Deployment.apps
        "data-science-pipelines-operator-controller-manager" is invalid:
        spec.selector: Invalid value:
        v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/name":"data-science-pipelines-operator"},
        MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable
```
```
Component reconciliation failed: Deployment.apps
        "odh-notebook-controller-manager" is invalid: spec.selector: Invalid
        value:
        v1.LabelSelector{MatchLabels:map[string]string{"app":"odh-notebook-controller",
        "app.kubernetes.io/part-of":"odh-notebook-controller",
        "component.opendatahub.io/name":"odh-notebook-controller",
        "kustomize.component":"odh-notebook-controller",
        "opendatahub.io/component":"true"},
        MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable
```
```
Component reconciliation failed: Deployment.apps
        "kserve-controller-manager" is invalid: spec.selector: Invalid value:
        v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/part-of":"kserve",
        "control-plane":"kserve-controller-manager",
        "controller-tools.k8s.io":"1.0"},
        MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable
```
```
Component reconciliation failed: Deployment.apps "rhods-dashboard" is
        invalid: spec.selector: Invalid value:
        v1.LabelSelector{MatchLabels:map[string]string{"app":"rhods-dashboard",
        "app.kubernetes.io/part-of":"rhods-dashboard",
        "deployment":"rhods-dashboard"},
        MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable
```

**Comment by shalberd** [comment]
ah, good to see. goes a little along the way as to why a default DSCI is even created in the first place after operator bundle install. I'd say listen to DSCI, but do not create it.
Also, that part of default dsco bundled with 2.4 leading to opendatahub namespace creation is not really good.

**Comment by zdtsw** [comment]
ref: https://github.com/opendatahub-io/opendatahub-operator/issues/706

**Comment by bdattoma** [comment]
@VaishnaviHire @zdtsw Is this considered "Closed" right?

**Comment by zdtsw** [comment]
> @VaishnaviHire @zdtsw Is this considered "Closed" right?

Yes

**Comment by VaishnaviHire** [comment]
Closing this as fixed in #706
